proxy.go

```diff
@@ -162,12 +163,96 @@ func (rp *reverseProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 	requestDuration.With(s.labels).Observe(since)
 }
 
+func executeWithRetry(
+	ctx context.Context,
+	s *scope,
+	retryNum int,
+	rp func(http.ResponseWriter, *http.Request),
+	rw ResponseWriterWithCode,
+	srw StatResponseWriter,
+	req *http.Request,
+	monitorDuration func(float64),
+	monitorRetryRequestInc func(),
+) (float64, error) {
+	// num of replicas should > 1,
+	// when the host is unavailable,
+	// it could make sure there might be another set of replicas contains all the data of clickhouse
+	startTime := time.Now()
+	var since float64
+	numReplicas := len(s.host.replica.cluster.replicas)
+	if numReplicas > 1 && retryNum <= numReplicas {
+		for i := 0; i <= retryNum; i++ {
+			rp(rw, req)
+
+			err := ctx.Err()
+			if err != nil {
+				since = time.Since(startTime).Seconds()
+
+				return since, err
+			}
+			srw.GetStatusCode(rw.StatusCode())
+			if rw.StatusCode() == http.StatusBadGateway {
+				if i == retryNum {
+					since = time.Since(startTime).Seconds()
+					monitorDuration(since)
+
+					s.host.penalize()
+					q := getQuerySnippet(req)
+					err1 := fmt.Errorf("%s: cannot reach %s; query: %q", s, s.host.addr.Host, q)
+					respondWith(srw, err1, srw.StatusCode())
+					return since, nil
+				} else {
+					h := s.host
+					log.Debugf("the invalid host is: %s", s.host.addr)
+					h.dec()
+					atomic.StoreUint32(&h.active, uint32(0))
+					s.host = h.replica.cluster.getHost()
+					if s.host.active == 1 {
+						req.URL.Host = s.host.addr.Host
+						req.URL.Scheme = s.host.addr.Scheme
+						log.Debugf("the valid host is: %s", s.host.addr)
+					}
+					monitorRetryRequestInc()
```

Comment:
This is going to increment the counter for newly selected host. It should be incremented before line 209. Given that, you'd rather accept labels as parameters to this method turning the signature to:
```
monitorRetryRequestInc func(prometheus.Labels)
```